src/codegen/statements.rs

```diff
@@ -946,31 +950,53 @@ fn destructure(
                 );
             }
             DestructureField::Expression(left) => {
-                // the resolver did not cast the expression
-                let loc = left.loc();
-
-                let expr = cast(
-                    &loc,
-                    right,
-                    left.ty().deref_any(),
-                    true,
-                    ns,
-                    &mut Vec::new(),
-                )
-                .expect("sema should have checked cast");
-                // casts to StorageLoad generate LoadStorage instructions
-                let expr = expression(&expr, cfg, contract_no, Some(func), ns, vartab, opt);
+                let expr = destructure_load(
+                    &left.loc(),
+                    &right.cast(left.ty().deref_any(), ns),
+                    cfg,
+                    vartab,
+                );
 
                 if should_remove_assignment(ns, left, func, opt) {
                     continue;
                 }
 
-                assign_single(left, &expr, cfg, contract_no, Some(func), ns, vartab, opt);
+                assign_single(left, expr, cfg, contract_no, Some(func), ns, vartab, opt);
             }
         }
     }
 }
 
+/// If the destructure statement returns a storage reference, we must load the value from storage
```

Comment:
This needs explaining. Why are the casts not in the ast for example.

Reply:
There is no cast in this function.

Reply:
I updated the comment. What is basically happening at this function is that we are adding a Load instruction to the CFG, given that we only have a reference to variable and not its value (this is not a cast).